File: src/utils/server_state.py
"""
Read/write .slbp-server.json in the project root.

This file stores runtime port assignments so that `slbp ui open` can discover
the URL without needing the ports to be predetermined.
"""

from __future__ import annotations

import json
import os
from pathlib import Path

import httpx

_PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent
_STATE_FILE = _PROJECT_ROOT / ".slbp-server.json"


def write_state(flask_port: int, ui_port: int, proxy_port: int) -> None:
    """Write port assignments (and this process's pid) to .slbp-server.json."""
    _STATE_FILE.write_text(
        json.dumps(
            {
                "proxy_port": proxy_port,
                "flask_port": flask_port,
                "ui_port": ui_port,
                "pid": os.getpid(),
            },
            indent=2,
        )
    )


def read_state() -> dict | None:
    """Return the state dict, or None if the file does not exist."""
    if not _STATE_FILE.exists():
        return None
    try:
        return json.loads(_STATE_FILE.read_text())
    except (json.JSONDecodeError, OSError):
        return None


def clear_state() -> None:
    """Remove the state file if it exists."""
    _STATE_FILE.unlink(missing_ok=True)


def get_running_server_state(timeout: float = 1.5, retries: int = 2) -> dict | None:
    """
    Return the state dict iff a server is actually reachable, else None.

    A present .slbp-server.json only means *some* run wrote it -- it's left
    behind by an unclean exit (e.g. taskkill) just as often as by a live
    server, so presence alone can't be trusted. This probes an existing
    backend route through the gateway to confirm the process is actually
    live before treating the recorded ports/pid as current.

    A false "not running" here is much costlier than a false "running" --
    the former can lead a caller to spawn a duplicate server stack, the
    latter just means a retry -- so a single slow/flaky probe isn't allowed
    to be the final word; retries only stop early on success.
    """
    state = read_state()
    if state is None:
        return None
    proxy_port = state.get("proxy_port")
    if not proxy_port:
        return None
    for _ in range(retries):
        try:
            r = httpx.get(
                f"http://127.0.0.1:{proxy_port}/api/session-defaults", timeout=timeout
            )
            if r.status_code == 200:
                return state
        except httpx.HTTPError:
            pass
    return None

File: desktop/src/main/serverState.ts
import fs from 'node:fs';
import path from 'node:path';

export interface ServerState {
  proxy_port: number;
  flask_port: number;
  ui_port: number;
  pid?: number;
}

const HEALTH_TIMEOUT_MS = 1500;

export function readServerState(repoRoot: string): ServerState | null {
  const stateFile = path.join(repoRoot, '.slbp-server.json');
  try {
    if (!fs.existsSync(stateFile)) return null;
    return JSON.parse(fs.readFileSync(stateFile, 'utf-8')) as ServerState;
  } catch {
    return null;
  }
}

/** Best-effort removal of .slbp-server.json, mirroring clear_state() on the Python side. */
export function clearServerState(repoRoot: string): void {
  const stateFile = path.join(repoRoot, '.slbp-server.json');
  try {
    fs.unlinkSync(stateFile);
  } catch {
    // Already gone, or never existed -- fine either way.
  }
}

/**
 * Mirrors src/utils/server_state.py's get_running_server_state(): a present
 * .slbp-server.json only means *some* run wrote it (an unclean exit leaves
 * it behind just as often as a live server does), so presence alone isn't
 * trusted -- an existing backend route is probed through the recorded proxy
 * port to confirm the process is actually live.
 *
 * A false "not running" is far costlier than a false "running" here (it can
 * lead to spawning a duplicate server stack), so a single slow/flaky probe
 * doesn't get the final word -- retries only stop early on success.
 */
export async function checkServerRunning(
  repoRoot: string,
  retries = 2,
): Promise<ServerState | null> {
  const state = readServerState(repoRoot);
  if (!state || !state.proxy_port) return null;
  for (let i = 0; i < retries; i++) {
    try {
      const response = await fetch(`http://127.0.0.1:${state.proxy_port}/api/session-defaults`, {
        signal: AbortSignal.timeout(HEALTH_TIMEOUT_MS),
      });
      if (response.ok) return state;
    } catch {
      // not reachable this attempt
    }
  }
  return null;
}
